Upscaling Master Branch

ColAngel opened the pull request:
- Added upscaling to the master branch. 
- Changed code to match the formatting.
- Added simple (for now) error handling for the case some parameters are invalid.
- Added additional comments in shader and in some functions

I tried to go along with your recommendations. I hope the code match them at least in most cases. 
Unfortunatelly i ll not have an access to my computer in the next week. So i can not do any changes in that time. However i can respond to the comment here. 
I also was able to test different wrong usage scenarios. If upscale mode is on and run command is not, you will get black screen with sound :-). I ll add the note about this to the ini file.
The other way around works, but causes strange behavior sometimes , that i can not explain in the moment.
However no game crashes occured.

bo3b commented on Dependencies/d3dx.ini:
```diff
@@ -768,10 +843,10 @@ ps-t100 = stereo2mono bb
-;ps-t125 = StereoParams
-;vs-t120 = IniParams
-;ps-t120 = IniParams
+vs-t125 = StereoParams
```
These need to be commented out for top of tree. Might be from the original file you started with, this is a recent change.

I'll put these back when I edit the file.

bo3b commented on DirectX11/DLLMainHook.cpp:
```diff
@@ -228,12 +228,30 @@ static BOOL WINAPI Hooked_SetWindowPos(
 {
-	if (G->SCREEN_FULLSCREEN == 2) {
+	/*
+		TODO: What about this: (dont know how to test it)
```
For this, I'm not sure how the hooks will impact your code.

I think it is OK to just go with this, and see if we run into problems.  People using it in the field will report back as they try to use on games that require the hooks.  MGSV is the main game this was created for.

This code suggests it won't work for The Witness, but I'm sure someone will try it.

ColAngel commented:
Ok, thanks for reviewing, and some corrections. The hook of setwindowpos is necessary for the upscaling. Maybe the withness may still work. It depends on what the game does maybe it calls resizetarget and setposition like the witcher 3 does or something similar. (should try upscaling =2)

bo3b commented on DirectX11/HackerDXGI.cpp:
```diff
@@ -1947,6 +2148,26 @@ STDMETHODIMP HackerDXGISwapChain::GetDesc(THIS_
 	{
+		if (pDesc)
+		{
+			//TODO: not sure whether the upscaled resolution or game resolution should be returned
```
I'd follow the lead of the Direct Mode code, and return the forced resolution.  I'm also not sure if that is correct.

The problem is that our runtime environment is unbelievably complex, and includes crazy stuff like other injectors (ENB, Reshade), Steam overlay, Origin overlay, MSI Afterburner, and on and on.  

If we return the original resolution when one of those asks for the description, they'd get the wrong size, and draw their overlay wrong. 

No idea if that actually happens or not, just observing some of the edge cases we can run into.  We can revisit it if it turns out to be a problem.

ColAngel replied:
Yeah, i think it is the best way. Steam and uplay overlays seems to work correctly

bo3b commented on DirectX11/IniHandler.cpp:
```diff
@@ -1218,6 +1221,63 @@ static void EnumerateCustomShaderSections(IniSections &sections, wchar_t *iniFil
+
+	memset(desc, 0, sizeof(D3D11_SAMPLER_DESC));
+
+	//TODO: do not really understand the difference between normal and comparison filter 
```
Can't help here.  I'm a good code guy, but I don't understand a lot of the nuances of 3D and graphic programming.

Might be worth waiting for some feedback once people start to use and make suggestions.  Other people may have better ideas for best way to do sampling or ways to improve the image.

ColAngel replied:
After few test with different configs i find the current configuration produces the best quality. Tried witcher with 4k interlacing (game res full HD vs 2K). Both looks very good, and almost the same (fullhd) is a bit blurry, but maybe i can add a deblurr filter or somerhing like that. However this is a very small drawback for constant 60 fps :-)

bo3b replied:
Sounds good.  BTW, does Upscaling work with Reshade?  I can guarantee that someone will ask about supporting both at the same time.  They will possibly want to use Reshade's sharpening or colorizing.

ColAngel replied:
Dont know the reshade tool at all. But if it does not mess with screen resolution and works with sbs shader,  it should work.

ColAngel commented:
"For this, I'm not sure how the hooks will impact your code." what hooks do you exactly mean? Is there another winapi hooks i ve not noticed? If so ill need to take a look at them. 
However the old path for forcing resolution is deactivated as soon as the upscaling is on.

ColAngel commented on DirectX11/HackerDXGI.cpp:
```diff
@@ -167,6 +167,104 @@ HackerDXGISwapChain::HackerDXGISwapChain(IDXGISwapChain *pSwapChain, HackerDevic
 	}
 }
 
+HackerDXGISwapChain::HackerDXGISwapChain(IDXGISwapChain *pSwapChain, HackerDevice *pDevice, HackerContext *pContext, const DXGI_SWAP_CHAIN_DESC* fake_swap_chain_desc, UINT new_width, UINT new_height)
```
Sure, you are right. That would looks much better. I guess it should work, we only need to declare all relevant hackerswapchain members as virtual. Ill take a look at it as soon as i have the access to my pc.

bo3b replied:
Cool, sounds good.  For the HackerDXGISwapChain members, you won't need them to be virtual, because typically they already have an implementation.  You can just use the same routine signature and it will be overridden.  It looks like the key one will be the destructor.

(Even stuff I did not want to override needed to be overridden because of the funky Com object approach.  So I kinda-sorta convert all the Com objects we use into actual C++ objects.)

In CreateSwapChain, you'll create the proper version directly, or fall back to standard object, based on d3dx.ini requirements.

<br>
If possible, the best syntax will be to use the follow-on constructor style I show there.  <br>
"... : HackerDXGISwapChain(pSwapChain, pDevice, pContext)".  <br><br>

If something can't be rearranged or it doesn't make sense, you can always call the superclass directly with HackerDXGISwapChain::XX.

ColAngel commented on DirectX11/HackerDXGI.cpp:
```diff
@@ -167,6 +167,104 @@ HackerDXGISwapChain::HackerDXGISwapChain(IDXGISwapChain *pSwapChain, HackerDevic
+}
+
+
 HackerDXGISwapChain1::HackerDXGISwapChain1(IDXGISwapChain1 *pSwapChain1, HackerDevice *pDevice, HackerContext *pContext)
```
Yeah, i was gonna ask you, whether the other paths are used at all. All my games that i have tested use normal swapchain object. With the time  Ill add the upscaling code in those functions too. Maybe you can list some games besides dishonored that are using another swpchain creation path, so i can test the updates directly.

ColAngel replied:
Heh, first read then write (its for me) just forget about sentence.

ColAngel commented on DirectX11/HackerDXGI.cpp:
```diff
@@ -813,7 +909,73 @@ STDMETHODIMP HackerDXGIFactory::CreateSwapChain(THIS_
 	}
 
-	HRESULT hr = mOrigFactory->CreateSwapChain(origDevice, pDesc, ppSwapChain);
+	HRESULT hr;
```
Yeah,with the new class it should be possible , dont see why not

ColAngel commented on DirectX11/HackerDXGI.cpp:
```diff
@@ -813,7 +909,73 @@ STDMETHODIMP HackerDXGIFactory::CreateSwapChain(THIS_
+		// TODO: testing testing testing
+		// not sure if the upscaling will work that way
+		// maybe i should better disable upscaling if 3dv direct mode is set
+		if (G->gForceStereo == 2)
```
Does rise of tombraider uses direct mode? Do you know that? Or maybe another game you know? I would test it.

bo3b replied:
No, no game presently uses the Direct Mode that I added.  It's currently specific to a Unity plugin that sgsrules is creating.  First game we expect to see use it is Inside. 

The TombRaider uses their built-in 3D.  Similar to GTA5 and Trine.  Those probably actually do use the NVidia Direct Mode in the driver, but won't need this code because they already handle it.  

Pretty sure the SBS shader will work on TombRaider as it stands, no normal 3Dmigoto code, just the SBS shader at the end.  Seems like Upscaling should also be fine.

ColAngel commented on DirectX11/HackerDXGI.cpp:
```diff
@@ -1891,6 +2061,38 @@ STDMETHODIMP HackerDXGISwapChain::SetFullscreenState(THIS_
+
+		// dont call setfullscreenstate again to avoid starting mode switching and flooding winproc with unnecessary messages
+		// can disable fullscreen mode somehow
+		if (fst && Fullscreen)
```
The full_screen flag is ignored as soon as upscaling is on. The fullscreen check is necessary to avoid state changes in direct x. Somehow if fullscreen is on and you call the game calls the function again the screen goes to windowed mode on some games(witcher). I could change it the way full_screen flag is used, but it would do the whole configuration more complicated. Moreover i have designed the upscaling to be always used with the fullscreen. Thats why i force migoto to ignore old resizing functionality if upscaling is on. This way the user desides whether to use "old" resizing or upscaling but in fullscreen.

bo3b replied:
OK.  If we run into games where it proves to be a problem, we can revisit it then.

bo3b commented on DirectX11/HackerDXGI.cpp:
```diff
@@ -167,6 +167,104 @@ HackerDXGISwapChain::HackerDXGISwapChain(IDXGISwapChain *pSwapChain, HackerDevic
+	{
+		LogInfo(" *** Failed to create faked swap chain buffer! \n");
+		mFakeBackBuffer = nullptr;
+		throw;
```
Hmmm... Is 'throw' the only option here?  That will just crash silently in a lot of games, with no way for them to know what happened.

If the swap chain texture fails, wouldn't it make more sense to fall back to normal behavior instead?  Probably disable the upscaling and just return?

Our default for configuration errors is to make a double-boop sound (found in Utils.)  But keep going.  That way the user knows something is wrong, but it's not a fatal error.

ColAngel replied:
The throw was already there. Did not added it as i wanted to keep staandard behavour. But it seemed strange to me too. Since the code is in constructor i d add a custom exception and catch it within the createswapchain. It is the cleanest solution. However cant think the case why the texture creation may fail.

bo3b replied:
Interesting.  Thanks for the detail.  I'll research when that was added and see if I can understand why.

bo3b replied:
OK, took a look for our code history, but don't find any prior usage of throw for that.  Went ahead and took a look out our other uses of throw just to see if we might have some holes, but it looks mostly OK.  

However, for this scenario, it does seem like throw is the only option for error handling there.  For standard C++ constructors, there is no error return, and the error handling is just to throw and exception.

In this case, while we could continue running in the scenarios where the texture or 2nd bb failed, it doesn't seem worth that much effort.  (to separate out that into a different initializer function)

So I'd say throw is probably OK here as the way to handle possible DX11 errors.  If this proves to be a problem, we can change it later, but for now, I'd say go with the throw.  

<br>
Add a flush for the log file before the throw, so that the log is definitive on what happened.  

And probably add a bit more info to the LogInfo to explain what happened so the user knows what to try, and that the problem is related to Upscaling.

ColAngel replied:
Im sorry. My bad. It seems that it easlate in the night as i addedthis shit. Try and catch block was there. My brain f** with me. Sorry. However, i will add proper error handling. I ll take time for it. :)

ColAngel commented:
As promised, moved upscaling functionality to the HackerUpscaleDXGISwapChain class and removed throw by adding exception class and catching it in the create swap chain function.
I ve forced the fail case to check the fallback strategy (disabling the upscaling). It works flawessly. The run = CustomShaderUpscale) causes no errors or issues even if upscaling mode is disabled

bo3b commented:
Great!  Let me take a look tomorrow or Friday, and I'll likely merge it in.

Thanks.

bo3b commented:
Sorry for the delay on reviewing here.

This looks pretty good, I'm pretty happy with the change to having the subclass of HackerDXGIUpscalingSwapChain.  This gets all the upscaling code into one code path, without having to mentally switch back and forth from standard to upscaling case in the HackerDXGISwapChain case.  Maybe a little bit of duplicated code, but that seems like a fair price for the clarity. 

I looked for places where we could call out to the superclass, but nothing really makes sense there.  It's more of a replacement than an extension, so I think that is all OK.


I've done the structural review, and I think this approach is good.  No particular changes to structure.

I will make some minor direct edits for code style, just for consistency.  Once I finish these minor edits, I'll probably go ahead and merge it in.  Should be done today.

Thanks for your effort on this.

ColAngel commented:
Thats very nice! I m a kind of happy about this :-)

I will try to respect code stylistics even more in the future. It is just the way im used to write code.  I just didn pay attention to it. 
About "#include ...crc32.h" - line it was not plannet to commit it that way, but you maybe need to check project settigs. The code does not compile on a clean system after checkout because of the line. 
(maybe you have the library path as some kind enviroment variable on your system thats why you have not noticed it yet). I did not want to mess around with project settings. 
I will go ahead and add the upscaling code for the other create swap chain paths as soon as i can find time for it. 
Thank you too!

bo3b commented on DirectX11/HackerDXGI.cpp:
```diff
@@ -2062,6 +2164,296 @@ STDMETHODIMP HackerDXGISwapChain::GetLastPresentCount(THIS_
+		mFakeSwapChain->Release();
+	if (mFakeBackBuffer)
+		mFakeBackBuffer->Release();
+}
```
Shouldn't this also call through to the superclass?  Just in case anyone adds something to that class?

Probably not virtual, letting it go to default seems better to me.

bo3b commented on DirectX11/HackerDXGI.cpp:
```diff
@@ -2062,6 +2164,296 @@ STDMETHODIMP HackerDXGISwapChain::GetLastPresentCount(THIS_
+		Anyway need to consider new code for the upscaling later
+	*/
+
+	// Some games like Witcher seems to drop fullscreen everytime the resizetarget is called (original one)
```
Witcher? DX9 game. If you mean Witcher3, please use that.  It's very confusing otherwise.  More keystrokes are always better for clarity.  Or to be more accurate, too few keystrokes adds confusion and extra mental work.

I really appreciate your 'why' comments.  One of the other big value comments is a 'theory of operation' comment.  Maybe add that to the HackerDXGIUpscalingSwapChain object. Like the comments from the original Issue. No rush.

bo3b commented:
Looks good, big thanks for setting this up and working through the review with me.  

I try not to be too dogmatic about the style stuff, but it does help smooth things out mentally when reading code.  Our code style is a bit of a mess already, just trying to keep it from getting worse.

For crc-32 thing, I'll take a look at that.  I just set up a brand new project on a Win10 box, and it has been working for me.  Is it possible that was for 2015 builds?

ColAngel commented:
"..just trying to keep it from getting worse" - it is completely understandable and must be done that way - one project one code style :-)
About crc i have the same issue on both builds 2015 and 2013 on win 10 (tested it on two different PCs). Maybe the location settings of windows could cause the problem (I m using german version of Window 10) 
I lltry to find out what exactly causes the problem on my system and let you know then.

ColAngel commented:
And thank you for putting your trust in me and giving me write access to the repository :-) I ll try my best :-)

bo3b commented:
Sounds good.  When you think it is ready, I'd say go ahead and merge it into the main branch.  Use the normal MergePullRequest, as keeping the history is valuable.

For making changes, I'm sure you'll do great.  As you make changes, I'll review them as my time permits, and make suggestions.    

If you get to where you want to make release builds, we use the Publish.bat script to automate this and avoid making mistakes.  For this you need to have the git installed for command line access, as it's used to automatically commit the version.h change.  No problem if you want me to keep doing these for the time being.

<br><br>
For the crc32c header problem, I'm not sure what's up.  When I built it here on my Win10 machine, it builds with no errors and no need for further changes.

I checked the project file, and the include directories also include the $(SolutionDir)crc32c-hw-1.0.5 directory. 

Look for the properties of DX11 chunk, and check the VC++ Directories section.  That should have the crc32c for all four build targets.

Not sure what's happening. Could be my version of 2013 is the full version, and it's a free version difference. Should be a relative path from the solution directory, so language and location shouldn't matter.  I guess the most likely answer is the different base language of German, but I don't have a good idea why.

llyzski commented:
I also do not have problem with crc32c.h include. Using VS2013 professional on win10.

ColAngel commented:
Its very strange :-). Have also the ultimate edition 2013 and enterprise 2015. It seemst the german loc indeed causes the issue. However i ll try to find solution for it as soon as i m back home again.   
I ll try to merge the upscaling code to the main branch (hope will not mess something up) Is there a possibility to revert changes to the previos revisions? (just in case something goes wrong)
To the release build: do it better yourselft to avoid any problems at least this time. I ll take a look at it one day.

bo3b commented:
Ah! Didn't realize that it was d3dcompiler_46. No, the problem is the d3dcompiler_46.dll.  Don't build that, it's at least partly obsolete, and you definitely do not want our version to replace a game version. The only value of that currently is in the Batman Arkham Knight fix, where it's a quick hack to get 3Dmigoto to load earlier than it otherwise would.

Probably that d3dcompiler_46 code has not been updated since the crc32c code was added.

There are several projects that are obsolete, or out of date.  DirectX9 doesn't work, DirectX10 was experimental and is probably broken.  D3DCompiler_39 through _43 are obsolete.
<br>

The only one you want to build is right-click DirectX11 and do Build.  You'll get the right pieces because of the project dependencies, which you can find in Configuration Manager.

<br>

When you are ready to do the merge, go ahead and just hit the GitHub MergePullRequest button, and it will create an automatic merge.  This will keep the history of changes too, which is best.  
 
Don't worry about it possibly breaking anything.  We will always have occasions where something goes astray and get the "works on my machine" problem.  But with full history we can either fix it, or revert it as we need.   Git is actually too flexible in this regard, it can get complicated fast, but we can always revert if we need to.

<br>

Best strategy for those is to checkout the latest top of tree, right before you plan to merge.  Or resync your fork with the top of tree, so that you are sure they are the same.  Then rebuild your fork with your changes and make sure it builds and runs like you expect.  

Then as long as top of tree on master did not change, you can just hit the green button and it will create an automatic merge.  Put in a good comment so it's clear what happened, otherwise it will be some generic 'merged from fork' message.

llyzski commented on Dependencies/ShaderFixes/upscalevs.hlsl:
```diff
@@ -0,0 +1,31 @@
+			break;
+		default:
+			pos.xy = 0;
+			break;
```
There is a compile warning here: warning X4000: use of potentially uninitialized variable (texcoord)

texcoord must be set even it is not used.

ColAngel replied:
that not the problem in this case!  But the code that invokes the shader uses only four vertices ==> it is not possible to get into situation where the texture coordinates are not valid!
But to keep the code clean i ve fixed it already (Thx for the hint)

llyzski replied:
Yes it was not a logic problem, but if you don't fix HLSL warnings 3Dmigoto will "beep beep" during startup. :)

bo3b replied:
Good catch.  As a general rule, we try to fix all warnings. DarkStarSword and I made a lot of changes to clean up all the warnings, and will also fix complaints from static analysis. (even though a lot of those are overly strict)

ColAngel commented:
@bob: Merged everithing together and tested on my system! After small fix it works now! you can create a new release build if you want :-)

bo3b commented:
Thanks for double checking this, figured I'd break something with those offline edits.  

I'll probably make a new build later today.

llyzski commented:
@bo3b 

Ah, please also take a look at my pending PRs before you make a new build. :)
